src/webui/api/torrentscontroller.cpp

```diff
@@ -1041,3 +1041,112 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->addTag(tagTrimmed) && !torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
+                    {"tag", tagTrimmed},
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to add tag"}
```

Comment:
You don't need to call it "errorMessage", since it's field of "error" object. Just "message".